sdk/core/core/src/az_http_request_builder.c

```diff
@@ -3,48 +3,201 @@
+  uint8_t * const buffer_end = buffer.begin + buffer.size;
+  uint8_t * const headers_start_uint8ptr = (uint8_t *)headers_start;
+
+  size_t const max_headers = (buffer_end - headers_start_uint8ptr) / sizeof(az_pair);
+
+  // If there's enough space in the buffer, return the max number of elements it can fit (capped to
+  // uint16_t's max)
+  return buffer_end > headers_start_uint8ptr
+      ? (max_headers > ~(uint16_t)0 ? ~(uint16_t)0 : (uint16_t)max_headers)
+      : 0;
+}
+
 az_result az_http_request_builder_init(
     az_http_request_builder * const p_hrb,
     az_span const buffer,
-    int16_t const max_url_size,
+    uint16_t const max_url_size,
     az_const_span const method_verb,
     az_const_span const initial_url) {
-  (void)p_hrb;
-  (void)buffer;
-  (void)max_url_size;
-  (void)method_verb;
-  (void)initial_url;
+  AZ_CONTRACT_ARG_NOT_NULL(p_hrb);
+
+  if (!az_span_is_valid(buffer) || !az_const_span_is_valid(method_verb)
+      || !az_const_span_is_valid(initial_url) || max_url_size < initial_url.size) {
+    return AZ_ERROR_ARG;
+  }
+
+  if (buffer.size < max_url_size) {
+    return AZ_ERROR_BUFFER_OVERFLOW;
+  }
+
+  AZ_RETURN_IF_FAILED(az_span_set(buffer, '\0')); // zero the buffer; we don't have to do this
+
+  az_span uri_buf = { 0, 0 };
+  AZ_RETURN_IF_FAILED(az_span_copy( // copy URL to buffer
+      (az_span){ .begin = buffer.begin, .size = max_url_size },
+      initial_url,
+      &uri_buf));
+
+  az_pair * const headers_start = get_headers_start(buffer, max_url_size);
+  uint16_t const max_headers = get_headers_max(buffer, headers_start);
+
+  *p_hrb = (az_http_request_builder){ .buffer = buffer,
+                                      .method_verb = method_verb,
+                                      .url = uri_buf,
+                                      .body = { .begin = NULL, .size = 0 },
+                                      .max_url_size = max_url_size,
+                                      .max_headers = max_headers,
+                                      .retry_headers_start = max_headers,
+                                      .headers_end = 0 };
+
   return AZ_OK;
 }
 
 az_result az_http_request_builder_set_query_parameter(
     az_http_request_builder * const p_hrb,
     az_const_span const name,
     az_const_span const value) {
-  (void)p_hrb;
-  (void)name;
-  (void)value;
+  AZ_CONTRACT_ARG_NOT_NULL(p_hrb);
+
+  if (!az_const_span_is_valid(name) || !az_const_span_is_valid(value)
+      || name.size == 0 // name can't be empty
+  ) {
+    return AZ_ERROR_ARG;
+  }
+
+  size_t new_url_size;
+  bool const has_value = value.size > 0;
+  {
+    az_span new_url_span = { .begin = p_hrb->url.begin };
+    size_t const extra_chars_size = sizeof((uint8_t)'?') + (has_value ? sizeof((uint8_t)'=') : 0);
+    size_t const name_and_value_size = name.size + value.size;
+    size_t const appended_size = name_and_value_size + extra_chars_size;
```

Comment:
If the '?' is already in the URL, then you don't need to add it.
Maybe the easiest approach here is to make sure that the initial URL has a ? in it when copied to the URL part of the buffer. Then this code assumes it's always there and never has to add it. And, to improve perf, we could add a field to the hrb struct that has the pointer to the '?' and then this method could always start from this pointer and never have to find the '?' ever again.

Then, when the final URL is pulled out, if it terminates with a '?' (with no query parameters after it), we can return the span when the end excluding the trailing '?'

Reply:
Victor also had similar proposal - to store boolean whether the URL has "?" already or not, then we need to scan the url for "?" during every parameter set.

To my taste, I'd leave this unoptimized. I don't think this is ever a perf bottleneck.
If we add "?" during init, we still have to check whether the initial URL already has parameters.
Plus, if we ever have remove_parameter() functionality, it now becomes harder to maintain consistency, and we put more and more assumptions and external state.

With my approach, I don't store any state, and I don't have to own all the edits to the URL. I can write a set of function to deal with URLs, and they would work with any az_span whose contents happens to be a URL at a time.

I think we can optimize my method, however. Instead of looking for "?", I can scan for either "?" or "&". This way, we have to scan even less. If we find either "?" or "&", we append next parameter with "&", otherwise we use "?".

Reply:
I am keeping it as is, but I added optimization to reverse search for either '?' or '&', until the first occurrence.